pycozmo/conn.py

```diff
@@ -158,19 +168,35 @@ def run(self) -> None:
             except Empty:
                 continue
 
+            raw_frames = []
+
             # Construct frame
             if isinstance(pkt, protocol_encoder.Ping):
-                frame = Frame(protocol_declaration.FrameType.PING, 0, 0, self.last_ack, [pkt])
+                raw_frames.append(
+                    Frame(protocol_declaration.FrameType.PING, 0, 0, self.last_ack, [pkt])
+                    .to_bytes())
             else:
-                seq = self.window.put(pkt)
-                if pkt.type == PacketType.COMMAND:
-                    frame = Frame(protocol_declaration.FrameType.ENGINE_ACT, seq, seq, self.last_ack, [pkt])
-                else:
-                    frame = Frame(protocol_declaration.FrameType.ENGINE, seq, seq, self.last_ack, [pkt])
-            raw_frame = frame.to_bytes()
+                first_seq, pkts = self.window.put(pkt)
```

Comment:
Instead of modifying the return value of `put()`, why not accessing `expected_seq` and `window` directly from `SendThread`?

Reply:
I guess just to keep a bit of the previous implementation and keep it into one line, but I agree that it would be clearer to separate the put method from the access to the current window and sequence.